### DISCUSSION TITLE
[IMPROVEMENT] labeling nodes hosting replicas to enable podAffinity

### DISCUSSION THREAD

**Discussion by dberardo-com** [longhorn/longhorn#9741] [General]
i am using a volume with 2 replicas on a 3 node cluster, with best-effort data locality startegy and read write once policy.

there is one pod using the volume, which needs fast read and write to disk so that's why the "best-effort" locality.

this pod is currently "free-to-move" among all cluster's node, but because the underlying volume has a lot of data, it would be desirable the have the pod constrained to just the 2 nodes hosting the volume replicas, otherwise if the pod starts on a node that does not have a replica, a huge data transfer process needs to start.

i thought of using podAffinity on the Pod definition for this, but how can i target the nodes hosting the replicas, is there any label i can use?

**Comment by DamiaSan** [comment]
Have a look here https://longhorn.io/docs/1.7.2/advanced-resources/deploy/node-selector/

**Reply by dberardo-com** [reply]
Thanks for the reply, but this is not solving my use case. 

So the issue here is not assigning replicas to nodes, but assigning pods (preference) to nodes that have replicas on them, to avoid rebuilding the replicas on other nodes if not necessary.

Please let me know if I have not 'mmade myself clear with my explanation above and I will rephrase the description.

**Reply by dberardo-com** [reply]
I hope that makes my point more clear? Or should I describe it Better?

**Comment by DamiaSan** [comment]
cc @ChanYiLin

**Comment by ChanYiLin** [comment]
@dberardo-com 
Lets say you have 3 nodes
How about
- Set the tag, for example, `special` on the 2 Nodes in Longhorn
- Create the volume with the `"Node Tags": "special"`, so the replicas will be only created on these 2 Nodes
- Then use kubectl to set the label to the nodes with `"tag"="special"`
- Create the pod with the `nodeSelector: {"tag": "ssd"}}`

The concept here is to 
1. force the replicas to some set of the node
2. force the pods to same set of the node

We can only help on the first one, 
but for second one it is outside the scope of Longhorn, above is just one possible way.
You can search how to force pod to be created on some nodes.

**Reply by dberardo-com** [reply]
sure this would work, but is a workaround that cannot scale well, since i would need to add nodeselector to essentially all staetful pods and  nodes that should host replicas for those pods. In practice this would lead to a suboptimal resource usage and so i would definitely want to avoid this.

TBH i don't see any other solution to my issue as to have longhorn add some sort of label to nodes where there are replica. 

i am not expert of k8s so i dont know which mechanism regular volumes use in order to for pods to stick to them, i know they dont use labels or affinities, so i was wondering if longhorn could make use of the same kind of system in order to constraint pods to spawn only on nodes where replicas are available.


can you get my point ? was i clear enough? thanks

**Reply by ChanYiLin** [reply]
I know your point 
But the pod scheduling is outside of Longhorn's control
It is managed by K8s scheduling itself.
And the K8s scheduler don't consider the scenario you mentioned.
They only provide node-affinity/taint mechanism for you to place the pods on the nodes you want.
Again, Longhorn cant control users pods and it shouldn't.

**Reply by ChanYiLin** [reply]
Even if you use GCP/AWS/Azure K8s service with their native storage services
They don't have this kind of setting to guarantee you that your pod will be on the same node as the volume.
Especially I believe all kind of this cloud storage solutions are supported by the network protocol.
The volume can be even on the node in another data center.

Adding labels to the node and adding node-selector or node-affinity to your pods is the most standard solution to this and is encouraged by the K8s official docs.

**Reply by dberardo-com** [reply]
> They only provide node-affinity/taint mechanism for you to place the pods on the nodes you want

these are techniques i know of, but i am not sure how k8s handles pod <-> volume stickiness contraint (affinity?) for regular volumes, think hostPath, or local provisioner in k3s, etc. I was hoping there was some advanced hidden mechanism that could be used to force pods on the same nodes as a volume

> Again, Longhorn cant control users pods and it shouldn't.

i understand, but cloud providers charge big money for moving data around between data nodes, so this would make quite an impact on budgets (and performance too)

> Adding labels to the node and adding node-selector or node-affinity to your pods is the most standard solution to this and is encouraged by the K8s official docs.

i see ... is there any pod in the longhorn-system namespace that one could make use of in order to create a podAffinity for the workloads ? Is there any pod, like the instanceManager for example that can be used to trigger the desired mechanism? I have not checked what happens if a node goes down whose volume was attached to the workload. is the instanceManager spawning on a new node BEFORE the workload is scheduled or does it happen AFTER ? if it happens before, then perhaps one could use the instanceMAnager as the affinity source

**Reply by ChanYiLin** [reply]
In short,  "k8s handles pod <-> volume stickiness contraint (affinity?)" doesn't exist in K8s

Pod is pod, managed by K8s and you
Volume is volume, managed by the third party, e.g. Longhorn, Ceph, GPC, AWS,...

As far as I know, hostPath only allows you to mount the path in the host to the pod on the same node. But it doesn't include forcing your pod to the node where you are going to provide the host path.
You still need to use node-selector to force the pod to the node you want, am I right?

For local-path-provisioner, it also don't include the user pods scheduling.
It only allows you to pre-defined the path in the host that you want it to be a hostPath
Then when the pod using that PVC is scheduled to one of the node, it will mount the pre-defined hostPath.
For example, if you have this config in the local-path-provisioner
```
  config.json: |-
        {
                "nodePathMap":[
                {
                        "node":"DEFAULT_PATH_FOR_NON_LISTED_NODES",
                        "paths":["/opt/local-path-provisioner"]
                },
                {
                        "node":"node1",
                        "paths":["/node1", "/data1"]
                },
                {
                        "node":"node2",
                        "paths":["/node2", "data2"]
                }
                ]
        }
```
> If one node is listed on the nodePathMap, the specified paths in paths will be used for provisioning.
If one node is listed but with paths set to [], the provisioner will refuse to provision on this node.
If more than one path was specified, the path would be chosen randomly when provisioning.

If your data is on node2 "/node2", you still need to specify the pod to be scheduled to the node2, so it can access the data in the path "/node2" on node2.

**Reply by ChanYiLin** [reply]
Instance-Manager is like a daemon-set which will be created on every nodes to manage the replica/engine processes on it.
When the node goes down, the instance-manager on that node won't be created on another node since it is 1:1 mapping.
In this case, replica on that node(instance-manager) goes down, the volume still works because it use network protocal and can access another replica data on another node(instance-manager).
If the node also contains YOUR pod, then your pod will be scheduled to another node BY K8S (if yours is a deployment) and Longhorn will  remount the volume to the new recreated pod and connect to other replicas.
That is the fundamental difference between using hostPath or local-path-provisioner. The whole IO system built on top of the network protocol allows the next-level resilience.

**Reply by ChanYiLin** [reply]
Again, volume scheduling is managed by Longhorn.
But pod scheduling is managed by K8s and You.

**Reply by ChanYiLin** [reply]
Moreover, volume is just a concept
For Longhorn, it is an iscsi initiator and target between engine and replicas on different node.
The IO from your pod goes through the engine and is dispatched to all the replicas and even wait until all replicas ack to ensure data consistency.
You cant say you want you pod to be as close as the volume to shorten the data path as long as you have extra replica on other nodes.

For OpenStack, the volume is a logical volume on different node as your VM, and it also uses the iscsi to dispatch the IO.
For Ceph, the volume is a on the RADOS layer developed by themselves which is a distributed storage system, your volume is divided by multiple blocks and stored by sharding on different nodes in the cluster. There is of course no pod-volume affinity in this case.

So, if IO performance is really a key metric for your application.
I would suggest you should looks more into the storage solution you are going to use, see how the IO works.
And do the real benchmark to see which one you can tolerate.

Resilience, replication and fault-tolerance is a tradeoff of the IO performance because you need to add more layer into the path.

**Comment by dberardo-com** [comment]
> Instance-Manager is like a daemon-set which will be created on every nodes to manage the replica/engine processes on it.
When the node goes down, the instance-manager on that node won't be created on another node since it is 1:1 mapping.

I see ... so there is no pod in longhorn-system which is specific to a volume, meaning if i have 200 volumes there would be 200 pods of that specific kind. so no, we can't use such kind of pod as a target of pod affinity then.

Does perhaps longhorn annotate CRDs or change CRD statuses when replicas are scheduled ? I am not sure if this could be one hint in the direction of a custom solution.

> You cant say you want you pod to be as close as the volume to shorten the data path as long as you have extra replica on other nodes.

well in general yes, but i am being very pragmatic here, since i know the volume is scheduled by longhorn then i know i can benefit from data-locality if the replica exists on the same node as the pod ... and i want to make use of this fact.

Anyway, i thank you for your explanation, i just wanted to make sure that my use case was not possible to achieve with the current features of longhorn and you confirmed it. I understand there is the need of custom solution to achieve what i am looking for, but i am not sure which part of k8s i should be addressing for it.

My initial bet would be using the descheduler (https://github.com/kubernetes-sigs/descheduler) or perhaps a custom controller. But since i want this solution to be generic and not related to any CRD perhaps controllers aren't the way to go. Is there any other k8s component i can make use of? is perhaps the admission hook something to look into ? any suggestion here?

this contribution by one github user goes in this direction: https://github.com/longhorn/longhorn/discussions/5486#discussioncomment-11482262

**Reply by ChanYiLin** [reply]
Ok so if it is okay for you to program your own component in k8s
Then one way is to write your own admission webhook just like the discussion you shared
Then whenever your pod is created, maybe recognized by the label, then in the webhook you first find where the replicas are and add the node-selector to the pod to force it to be created on the nodes you want.
But this would still requires you to pre label the k8s node, so you can use node-selector to do the job.
Another issue is that the replicas can be recreated on other nodes when some failures happen.
You will need to recreate your pod to reschedule it.
But that might interrupt your application when actually the volume itself is still working, does that really worth it, you can keep in mind.

**Reply by ChanYiLin** [reply]
The custom controller you proposed in the thread is also one of the possible ways.
But the controller controls the CRD, that means you have to wrap your pod up with a CRD
So the controller does the job
1. Watch the update event of the LH replicas and your own CRD
2. When creating your CRD/pod, it list all the replicas of the volume to see where to start the pod
3. When any replica is updated, you get the pod which uses the volume of the replica to see if it is still on the same node of one of the  replicas. If no, you delete and recreate the pod

**Reply by ChanYiLin** [reply]
BTW, may I know what kind of the application you are going to deployed?
It seems it is so IO sensitive.

Longhorn would require SSD disks.
And just like I said, the IO goes to all replicas and wait until all replicas ack. So you may also need high speed network or the bottleneck will still be on the cross node data transfer even if you put the pod on the same node of one of the replicas.

**Reply by ghost** [reply]
> And just like I said, the IO goes to all replicas and wait until all replicas ack

@ChanYiLin 
This is good to know, so are all replicas considered _live_ and have the most updated data? Is this level granularity of technical stuff documented anywhere? Sorry don't mean to sideline the conversation this is the first time I've seen that information

**Reply by dberardo-com** [reply]
> BTW, may I know what kind of the application you are going to deployed? It seems it is so IO sensitive.

it is not necessarily IO sensitive, it is more related to the amount of data to be transferred between nodes when replica gets recreated. Say i have 1 volume that is 1TB big, i would prefer the pod to co-live on the same node as the replica (best-effort locality) but i would like to avoid useless rebuilding of the replica on another node, cause that would cause 1TB data transfer each time ...

> And just like I said, the IO goes to all replicas and wait until all replicas ack. 

i am hoping that in the future longhorn enable asynchronous replication as well ;)

> @cob-web-corner  Is this level granularity of technical stuff documented anywhere? 

indeed i was asking this somewhere else in the community, as the replication details are not so well documented in this sense. i wasnt sure that longhorn uses synchronous replication, now i know ;)

**Reply by dberardo-com** [reply]
BTW @ChanYiLin  we are discussing an approach to handle this using the Longhorn Node CRD and a custom controller that sets labels on actual k8s Nodes:

> UPDATE: just found out that longhorn is updating the Nodes CRD with a value called scheduledReplicas. Since podaffinity do not work with CRD, i think one could use a custom controller that checks this field and then update the actual Node's labels with the values from this CRD. Then the nodeaffinity rule could work on the pod to be scheduled on those nodes which have replicas on them.
hope this makes sense.

feel free of letting us know if that makes sense

https://github.com/longhorn/longhorn/discussions/5486#discussioncomment-11484654

**Reply by ChanYiLin** [reply]
So there are two issues here
>  i would prefer the pod to co-live on the same node as the replica (best-effort locality) but i would like to avoid useless rebuilding of the replica on another node, cause that would cause 1TB data transfer each time ...
1. pod to co-live on the same node as the replica
2. avoid useless rebuilding of the replica on another node, cause that would cause 1TB data transfer each time

For 1, as discussed before, users' pods are out of Longhorn's control. So this part would require you to do the setting.
For 2, 
- Normally when the replica is created and scheduled on to a node, it won't rescheduled to another one. 
- Replica consists of a series of snapshot (image file) on the host disk. These image files hold the real data.
- Replica is a process on the host managed by the instance-manage pod
- The failure can be broken down into several cases 
     - If the replica process some how crashes, instance-manager will recreate the process and reuse the snapshot(image file)
     - If the node/disk goes down. Longhorn consider it as temporary outage, it won't recreate the replica on another node
- When the replica is recovered from the crash, Longhorn sync the data from healthy replica by copying the snapshot to the rebuit one
    - However, we have [Fast Replica Rebuild Enabled](https://longhorn.io/docs/1.7.2/advanced-resources/rebuilding/fast-replica-rebuild/) enabled by default, which allows Longhorn to only copy the necessary snapshot image => Less data transfer
- When the replica CR is created, it is already assigned to a node/disk.
- That is, only when the replica CRs count is less then the `volume.spec.replicaCount` will it create a new replica on a new node and requires full data transfer

By the way, Longhorn use Linux sparse file as the image file. So even the volume has 1TB, the real size of the volume will be the size of the data you really write into it. 

Ref
- How Longhorn works: https://longhorn.io/docs/1.7.2/concepts/#2-longhorn-volumes-and-primary-storage
- Longhorn at Kubecon: https://www.youtube.com/watch?v=ZZblyTjrEwA&

**Reply by ChanYiLin** [reply]
Oh wait, I see your problem here
> pod is currently "free-to-move" among all cluster's node, but because the underlying volume has a lot of data, it would be desirable the have the pod constrained to just the 2 nodes hosting the volume replicas, **otherwise if the pod starts on a node that does not have a replica, a huge data transfer process needs to start.**

**No, the replica won't be recreated on the node where your pod is on, Longhorn won't do that for you** , you don't need to worry about the data transfer in this case
And you don't need to reschedule the replicas as well. There is no point of doing it. It only introduces more overhead.
If the IO performance through the network with this architecture fits your requirement. Then everything is fine.
Most of the cloud native storage solutions all include these kind of network interface. You can't really say you want to schedule the pod to the volume as the "volume" is just a virtualization resource. K8s doesn't work in that way.

```
Pod -> Volume -> Engine -> [Replica1, Replica2]
```
The IO will be dispatched by the Engine to the Replicas. The replicas can be anywhere in the cluster.
We don't rescheduled the Replicas to another node when the pod is recreated to another node. We don't need to.

**Reply by ChanYiLin** [reply]
This longhorn architecture might answer your question. 
![Screenshot from 2024-12-07 22-22-05](https://github.com/user-attachments/assets/3bab9761-c6c9-46dc-8fa9-cb14e779f725)

But suggest you to watch through the conference talk to get more understanding.

**Reply by ChanYiLin** [reply]
> BTW @ChanYiLin we are discussing an approach to handle this using the Longhorn Node CRD and a custom controller that sets labels on actual k8s Nodes:
> 
> > UPDATE: just found out that longhorn is updating the Nodes CRD with a value called scheduledReplicas. Since podaffinity do not work with CRD, i think one could use a custom controller that checks this field and then update the actual Node's labels with the values from this CRD. Then the nodeaffinity rule could work on the pod to be scheduled on those nodes which have replicas on them.
> > hope this makes sense.
> 
> feel free of letting us know if that makes sense
> 
> [#5486 (reply in thread)](https://github.com/longhorn/longhorn/discussions/5486#discussioncomment-11484654)

This should work.
You have a set of replicas managed by Longhorn.
You have your pod.

To put them together. You have to fix one of them, then put another one close to it.
When volume is created, replicas are scheduled as well. So replicas are fixed.
Then yes, you can have your custom controller to put the pods close to the replicas.

**Reply by ghost** [reply]
@ChanYiLin Thanks for the learning material. Are there other in depth technical discussions/guides you would recommend that aren't in the official docs? Loving Longhorn so far it has passed a lot of my initial tests and benchmark needs.

**Reply by ChanYiLin** [reply]
@cob-web-corner 
I think most of the technical details has been included in the official docs.
The core volume/engine/replica design has been implemented since the very early version.
Maybe, you can take a look at all the settings. We provide large number of settings relating to the failure handling, scheduling, space utilization to fit users need.
And you may learn more about the details and limitations of Longhorn from there.
Thanks.

**Reply by ChanYiLin** [reply]
Hi @dberardo-com @cob-web-corner 
After reminded by the team members.
I may have some misunderstanding to your question here. As there is presumption,`best-effort` locality, working in your environment.
Yes, if you enable "best-effort" locality here, Longhorn will rebuild the replica on the node of your pod and delete the extra replica.
This indeed will introduce an extra data transfer cost, but, without any downtime. 

This is a trade-off of the IO performance and the underneath data transfer cost.
And yes, Longhorn automatically tries to rebuild a replica on the node where your pod is on (if there is no replica on that node). 
So one possible way of avoid the "data transfer cost" is to always starts your pod on the nodes where the replicas are already on.
And since the replicas are scheduled when the volume is created.
Yes, you can have a custom controller to control you pod as we discussed.
Or one more simpler way, is to have a webhook always patch your pod with the `node-selector` to force create the pod on those nodes where replicas are.

**Reply by ChanYiLin** [reply]
Hi @PhanLe1010 @c3y1huang 
Do you have any more comment for [this](https://github.com/longhorn/longhorn/discussions/9741#discussioncomment-11508983)

**Reply by ghost** [reply]
>  one more simpler way, is to have a webhook always patch your pod with the node-selector to force create the pod on those nodes where replicas are

@ChanYiLin Mutating webhook is definitely simpler agreed, this is how I'm rolling now. Works flawlessly. I would argue Longhorn should implement this, ~~I totally understand compromise and all of that but rebuilding the replica on the new node seems like putting the cart before the horse (allowing the pod to be placed willy-nilly and then chasing it with a replica rebuild)~~
ignore strike through, given how the storage network/access works this makes sense, there's a lot to consider here

**Reply by dberardo-com** [reply]
@ChanYiLin  as per my very initial message (https://github.com/longhorn/longhorn/discussions/9741#discussion-7400321) :D :

> i am using a volume with 2 replicas on a 3 node cluster, with **best-effort** data locality strategy and read write once policy

Never mind, i am glad that we cleared out the confusion.

As for the webhook ... i am not that advanced in k8s to know what this imply, that's why i came up with the controller idea. @cob-web-corner  feel free of providing a brief explanation of what needs to be done for the webhook, as i am missing the fundamentals :D 

Also, please notice:

* we should possibly make the node-labeling behavior optional for every volume, as when one has lot of volumes and replicas this might end up filling up nodes with lot of labels, which might not be a best practice
* @ChanYiLin  is it possible to label the nodes using the volume / PVC name ? i mean, in order for the user to properly create nodeAffinity rules, those rules need to be known in-advance to the user, so we cannot use replicas names or volume-internal IDs, but names coming from user-defined objects such as the PVC name

**Reply by ghost** [reply]
@dberardo-com If you can wait a little while I can build something for you

>we should possibly make the node-labeling behavior optional for every volume, as when one has lot of volumes and replicas this might end up filling up nodes with lot of labels, which might not be a best practice

There's no need to modify or label any resources. This can be done on the fly with the mutation webhook

**Reply by ghost** [reply]
@dberardo-com Did you still need this? I found an even simpler way of doing this with new additions to Kubernetes since 1.29 and custom scheduler

**Reply by dberardo-com** [reply]
Sounds great, sure, I still wish to test some solutions within the next weeks so this could be helpful

**Reply by ghost** [reply]
New method is pretty much:

- On pod definition name a custom scheduler `schedulerName: pvScheduler`. 
- That scheduler needs read access to PV bindings. 
- It can tell what PVC correlates to which PV because of the pod definition and information in the PV longhorn creates

It will leave the pod in pending state until the PVs are created. When they are finish it knows which nodes they were bound to. The scheduler then files a `binding` resource to bind pod to one of the nodes with PV.

Everytime the pod needs to be rescheduled it will use this scheduler. It will be able to do this stateless

**Reply by dberardo-com** [reply]
i see, i am not familiar with this concept either. do we need to create a custom scheduler ? is this a global or a namespaced resource? and how would the definition look like ?

and other then that: would then suffice to add the schedulerName attribute to all deployments/statefulSets requiring and thats it ? no need to modify also storageclasses or PVC ?

Also: why is k8s v 1.29 needed for that ? which feature is the new one ?

**Reply by dberardo-com** [reply]
@ghost in the incoming weeks I might find the time to test out some possibilities, would you feel like sharing a GitHub gist with some indications so I can comment directly in there and propose changes if needed?

Thanks for the collaboration

**Comment by yama6a** [comment]
Hit this too. Ended up skipping the labels entirely and using a mutating webhook instead, which looks up where the replicas actually are as the pod is created and adds a preferred nodeAffinity for those nodes. Nothing to maintain by hand and it just falls back to normal scheduling if neither node can take the pod: https://github.com/yama6a/longhorn-replica-affinity
